src/aspire/abinitio/commonline_c3_c4.py

```diff
@@ -0,0 +1,363 @@
+        # return vijs, viis
+        pass
+
+    #######################################
+    # Secondary Methods for Global J Sync #
+    #######################################
+
+    def _J_sync_power_method(self, vijs):
+        """
+        Calculate the leading eigenvector of the J-synchronization matrix
+        using the power method.
+
+        As the J-synchronization matrix is of size (N choose 2)x(N choose 2), we
+        use the power method to the compute the eigenvalues and eigenvectors,
+        while constructing the matrix on-the-fly.
+
+        :param vijs: nchoose2x3x3 array of estimates of relative orientation matrices.
+
+        :return: Array of length N-choose-2 where the i-th entry indicates if vijs[i]
+        should be J-conjugated or not to achieve global handedness consistency. This array
+        consists only of +1 and -1.
+        """
+
+        n_vijs = vijs.shape[0]
+        nchoose2 = (1 + np.sqrt(1 + 8 * n_vijs)) / 2
+        assert nchoose2 == int(nchoose2), "There must be n_ims-choose-2 vijs."
+        # assert n_eigs > 0, "n_eigs must be a positive integer."
+
+        epsilon = 1e-2
+        max_iters = 100
+
+        # Initialize candidate eigenvectors
+        vec = np.random.randn(n_vijs)
+        vec = vec / norm(vec)
+        dd = 1
+        itr = 0
+
+        # Power method iterations
+        while itr < max_iters and dd > epsilon:
+            itr += 1
+            vec_new = self._signs_times_v(vijs, vec)
+            # vec_new, eigenvalues = qr(vec_new)
+            vec_new = vec_new / norm(vec_new)
+            dd = norm(vec_new - vec)
+            vec = vec_new
+
+        logger.info(
+            f"Power method used {itr} iterations. Maximum iterations set to {max_iters}."
+        )
+
+        # We need only the signs of the eigenvector
+        J_sync = np.sign(vec)
+
+        return J_sync
+
+    def _signs_times_v(self, vijs, vec):
+        """
+        For each triplet of outer products vij, vjk, and vik, the associated elements of the "signs"
+        matrix are populated with +1 or -1 and multiplied by the corresponding elements of
+        the current candidate eigenvector supplied by the power method. The new candidate eigenvector
+        is updated for each triplet.
+
+        :param vijs: Nchoose2 x 3 x 3 array, where each 3x3 slice holds the outer product of vi and vj.
+
+        :param vec: The current candidate eigenvector of length Nchoose2 from the power method.
+
+        :return: New candidate eigenvector of length Nchoose2. The product of the signs matrix and vec.
+        """
+        # All pairs (i,j) and triplets (i,j,k) where i<j<k
+        indices = np.arange(self.n_ims)
+        pairs = [(i, j) for idx, i in enumerate(indices) for j in indices[idx + 1 :]]
+        trips = [
+            (i, j, k)
+            for idx, i in enumerate(indices)
+            for j in indices[idx + 1 :]
+            for k in indices[j + 1 :]
+        ]
```

Comment:
`trips = [(i, j, k) for i in range(self.n_ims) for j in range(self.n_ims) for k in range(self.n_ims) if i < j < k]`?

Reply:
Thanks. done.